cl/lib/elasticsearch_utils.py

```diff
@@ -176,22 +176,42 @@ async def build_more_like_this_query(related_ids: list[str]) -> Query:
     exclusions for specific opinion clusters.
     """
 
-    document_list = [{"_id": f"o_{id}"} for id in related_ids]
+    opinion_cluster_pairs = [
+        opinion_pair
+        for opinion_id in related_ids
+        if (
+            opinion_pair := await Opinion.objects.filter(pk=opinion_id)
+            .values("pk", "cluster_id")
+            .afirst()
+        )
+    ]
+    unique_clusters = {pair["cluster_id"] for pair in opinion_cluster_pairs}
+
+    document_list = [
+        {
+            "_id": f'o_{pair["pk"]}',
+            "routing": pair["cluster_id"],
+            # Important to match documents in the production cluster
+        }
+        for pair in opinion_cluster_pairs
+    ] or [
+        {"_id": f"o_{pk}"} for pk in related_ids
+    ]  # Fall back in case IDs are not found in DB.
```

Comment:
This fallback value creates the impression that `IDs` might exist in the Elasticsearch cluster but not in our database. While this method attempts to use these IDs when cluster pairs are unavailable, it raises concerns about scenarios where the `related_ids` list contains more `IDs` than matching clusters. In such cases, the unmatched IDs would be entirely ignored. Is this intended behavior, or should we consider including them without routing?

Reply:
Yeah, I got it your concern. The purpose of this fallback is only to prevent the MLT query from failing. If no documents are passed to the `like` parameter, the query will fail. So this is not meant to determine if any of the IDs that were not found in DB might return results. In the scenario where none of the opinions are found in the DB (if the user provided a wrong Opinion IDs), we simply pass the original IDs to the query to avoid the query to fail.
Since routing is required, and considering that the opinions IDs don’t exist in ES, even using the fallback list of IDs will return no results. In this case, I think that's better (returning no results) than display an error message due to a failed query. 

That’s why, for instance, if the user provides two IDs and only one of them is found in the database, it’s better to search for that ID alone and ignore the other one, as it won’t be found anyway due to the lack of routing.

I've improved the comment here to better explain its purpose.

Let me know what do you think.